### ISSUE TITLE
`writeFile` and `write(f, str)` skip null bytes on Windows

### ISSUE THREAD

**Issue by exelotl** [nim-lang/Nim#12315]
On Windows, when using `writeFile` or `write(file, string)` to output non-textual data, all null bytes will be removed in the output file.

The issue was introduced fairly recently by these two commits: https://github.com/nim-lang/Nim/commit/7cb31455ee949ee006e2f314daa74866cf37a74e, https://github.com/nim-lang/Nim/commit/742862b847646dff4562992cd502d3077b2c9474. It affects Nim 1.0.0, but not 0.20.2 or earlier.

### Example
```nim
import os
let str = "some nulls: \0\0\0 (three of them)"
writeFile("test.bin", str)
echo str.len
echo getFileSize("test.bin")
```

### Current Output
We can see that the lengths differ because the three null bytes have been removed:
```
31
28
```

### Expected Output
Lengths should be equal:
```
31
31
```

### Possible Solution
I've attempted to solve the issue in PR https://github.com/nim-lang/Nim/pull/12314, with an accompanying test.

However, my solution causes the [echo_unicode](https://github.com/nim-lang/Nim/blob/742862b847646dff4562992cd502d3077b2c9474/tests/system/techo_unicode.nim) test to fail, because that test *expects* the console output to not contain any nulls.

Is that really the correct behaviour? If so, @disruptek mentioned on IRC that we might need some distinction between console output and other file output.

**Comment by Araq** [comment]
Fix the test then. It was written with "text" in mind but it doesn't hurt to output binary zeros.

**Comment by exelotl** [comment]
I gave it a shot yesterday but it doesn't seem straightforward, if I put nulls in the expected output, the test still fails because the nulls somehow end up as newlines in the actual output. Is that the kind of issue you were encountering already before you switched to fprintf?

**Comment by Araq** [comment]
No, I switched to fprintf for Unicode support.

**Comment by pgkos** [comment]
The current implementation fails on Windows < 10, e.g.:
```nim
import strutils

echo "E" & "é".repeat(0x10000)
```
is truncated and rectangles are shown at the end.

**Comment by pgkos** [comment]
Maybe it would be possible to use `_write` for writing to console?

Windows console displays utf-8 characters correctly only if all the bytes of a utf-8 character are written using a single `_write` call. Unfortunately `fwrite` does split them into multiple calls, e.g.:
```c
const char *s = "é";

_write(1, s + 0, 1);
_write(1, s + 1, 1);  // bad - displays two rectangles

_write(1, s, 2);  // good - displays é

fwrite(s, 1, 2, stdout);  // bad - displays two rectangles
```

Pros:
* atomic
* can write null bytes, no need to search for them

Cons:
* not thread safe (custom locking needed in Nim stdlib)
* custom buffering implementation would be needed to ensure that utf-8 characters are not split inside multiple `_write` calls - on one of my Windows 7 machines, I can write at most 1024 bytes in one `_write` call, on another - 4096 bytes

e.g.:
```nim
if (file == stdout or file == stderr) and file.fileno().isatty():
  # use _write with custom locking and utf-8 buffering
else:
  # use fwrite
```

**Comment by exelotl** [comment]
I took another look at fixing techo_unicode, it seems like somehow testament doesn't like nulls in the output?

e.g. I change the test like so
```nim
discard """
  output: "ÄhmÖÜ\nabasdfdsmÄhmaИ\nИ\0nasty\0\0\0\0string\0\nA你好\nИ\0nasty\0\0\0\0stringA\0你好\nÖÜhmabasdfdsmÄhmaИ"
  disabled: "posix"
  joinable: "false"
"""
```

Then running `./koch test run system/techo_unicode.nim` gives the following result:

![image](https://user-images.githubusercontent.com/569607/66171521-c2b22780-e648-11e9-83a6-442dbb9995cc.png)

But if I run the test program directly, I do get the expected output:

![image](https://user-images.githubusercontent.com/569607/66171627-250b2800-e649-11e9-88d5-fa27ac52b16b.png)

I'm really not sure what's going on there.

In general I'm nervous about using fprintf for outputting raw bytes to a file. My workaround to add the nulls back seems to fix my projects, but from what I've read the printf family is not really intended for non-textual output. I'm no expert here though. @pgkos's `_write` solution sounds promising for more robust unicode output on older windows versions.

I didn't realise it was so simple to check if we're outputting to terminal. If fixing the test isn't straightforward, I guess for now we should just do that. Call `writeWindows` (old implementation) if isatty, otherwise call `writeBuffer` like normal. This would satisfy both tests. I'm just hoping there's no use case where raw binary output to the command line (including nulls) is needed.

**Comment by akavel** [comment]
How about combining the above solutions, i.e.:

```nim
proc write*(f: File, s: string) {.tags: [WriteIOEffect], benign.} =
  when defined(windows):
    if f.isatty:
      writeWindows(f, s, doRaise = true)
      return
  if writeBuffer(f, cstring(s), s.len) != s.len:
    raiseEIO("cannot write string to file")
```

Would that fix all the tests mentioned above, as well as the original issue? Or do some of them still break then?

**Comment by exelotl** [comment]
@akavel thanks, I've adjusted my PR because I think this is a better solution.

Unfortunately it doesn't fix `techo_unicode`, presumably because the test output doesn't go directly to stdout, therefore `isatty` returns false and we get the old fwrite behaviour.

This leads me to wonder, what is `techo_unicode` actually trying to test, and how do we test it? It seems like right now it wants to know "did the terminal interfere with the strings that I echoed?" which I don't see how to prevent, plus it's a hopeless question because testament is consuming the output. But really it should be asking "is the output visually correct to the user?" for which we'd need a real window that we can take screenshots from? That seems way outside the scope of these automated tests.
